src/manybody.jl

```diff
@@ -327,154 +316,101 @@ end
+function onebodyexpect_1(op::Operator, state)
+    b = basis(state)
+    occupations = b.occupations
+    S = length(b.onebodybasis)
+    buffer = allocate_buffer(occupations)
+    result = complex(0.0)
+    for i = 1:S, j = 1:S
+        value = op.data[i, j]
+        iszero(value) && continue
+        for (m, occ) in enumerate(occupations)
+            C = state_transition!(buffer, occ, i, j)
+            C === nothing && continue
+            n = state_index(occupations, buffer)
+            n === nothing && continue
+            result += C * value * get_value(state, m, n)
         end
     end
     result
 end
 
-function onebodyexpect_1(op::SparseOpPureType, state::Ket)
-    N = length(state.basis)
-    S = length(state.basis.onebodybasis)
-    result = complex(0.)
-    occupations = state.basis.occupations
-    M = op.data
-    @inbounds for colindex = 1:M.n
-        for i=M.colptr[colindex]:M.colptr[colindex+1]-1
-            row = M.rowval[i]
-            value = M.nzval[i]
-            for m=1:N, n=1:N
-                C = coefficient(occupations[m], occupations[n], row, colindex)
-                if C != 0.
-                    result += C*value*conj(state.data[m])*state.data[n]
-                end
-            end
+function onebodyexpect_1(op::SparseOpPureType, state)
+    b = basis(state)
+    occupations = b.occupations
+    buffer = allocate_buffer(occupations)
+    result = complex(0.0)
+    @inbounds for (row, column, value) in zip(findnz(op.data)...)
+        for (m, occ) in enumerate(occupations)
+            C = state_transition!(buffer, occ, row, column)
+            C === nothing && continue
+            n = state_index(occupations, buffer)
+            n === nothing && continue
+            result += C * value * get_value(state, m, n)
         end
     end
     result
 end
 
-function onebodyexpect_1(op::SparseOpPureType, state::Operator)
-    N = length(state.basis_l)
-    S = length(state.basis_l.onebodybasis)
-    result = complex(0.)
-    occupations = state.basis_l.occupations
-    M = op.data
-    @inbounds for colindex = 1:M.n
-        for i=M.colptr[colindex]:M.colptr[colindex+1]-1
-            row = M.rowval[i]
-            value = M.nzval[i]
-            for s=1:N, t=1:N
-                C = coefficient(occupations[s], occupations[t], row, colindex)
-                if C != 0.
-                    result += C*value*state.data[t,s]
-                end
-            end
-        end
+Base.@propagate_inbounds function state_transition!(buffer, occ_m, at_indices, a_indices)
+    any(==(0), (occ_m[m] for m in at_indices)) && return nothing
+    result = 1
+    copyto!(buffer, occ_m)
+    for i in at_indices
+        result *= buffer[i]
+        result == 0 && return nothing
+        buffer[i] -= 1
     end
-    result
-end
-
-
-"""
-Calculate the matrix element <{m}|at_1...at_n a_1...a_n|{n}>.
-"""
-Base.@propagate_inbounds function coefficient(occ_m, occ_n, at_indices, a_indices)
```

Comment:
given #147 , it seems some folks use this function even though it is internal. Do you mind keeping a version of it (a wrapper around your implementation) so that we do not break user code? I guess Hyrum's law strikes again https://www.hyrumslaw.com/

We should probably start keeping a changelog to make it easier for folks to notice when stuff like this happens.

Also, please go ahead and bump the version number to 0.4.18 - I will make a release immediately after merging this.

Reply:
Maybe also using `@deprecate` to warn people they should not be using this (they should never have started, but too late for that)

Reply:
Let me backtrack on this, I misunderstood the report in #147 . No need to add anything or many any changes. I will reread and merge this soon.